File: src/query/milvus-client.ts
import { MilvusClient, MetricType, DataType } from '@zilliz/milvus2-sdk-node';
import { SearchQuery, SearchResult, VectorMetadata } from '../types/index.js';
import { config } from '../utils/config.js';
import { logger } from '../utils/logger.js';

export class MilvusQueryClient {
  private client: MilvusClient;
  private isConnected = false;

  constructor() {
    const clientConfig: any = {
      address: `${config.milvusHost}:${config.milvusPort}`,
    };
    
    if (process.env.MILVUS_USER) {
      clientConfig.username = process.env.MILVUS_USER;
    }
    
    if (process.env.MILVUS_PASSWORD) {
      clientConfig.password = process.env.MILVUS_PASSWORD;
    }
    
    this.client = new MilvusClient(clientConfig);
  }

  async connect(): Promise<void> {
    try {
      // Test connection
      await this.client.checkHealth();
      
      // Use default database only
      await this.client.useDatabase({ db_name: 'default' });
      logger.info('Using default database');
      
      // Check if collection exists, create if not
      await this.ensureCollectionExists();
      
      this.isConnected = true;
      logger.info('Connected to Milvus successfully');
    } catch (error) {
      logger.error('Failed to connect to Milvus:', error);
      throw error;
    }
  }

  async disconnect(): Promise<void> {
    if (this.isConnected) {
      // Note: SDK doesn't have explicit disconnect, connection is managed automatically
      this.isConnected = false;
      logger.info('Disconnected from Milvus');
    }
  }

  async searchEmbedding(
    query: SearchQuery,
    queryVector: number[],
    topK: number = 10
  ): Promise<SearchResult[]> {
    if (!this.isConnected) {
      await this.connect();
    }

    try {
      const collectionName = this.getCollectionName(query.chunkType);
      
      // Build search params
      const searchParams = {
        collection_name: collectionName,
        vectors: [queryVector],
        search_params: {
          anns_field: 'vector',
          topk: topK,
          metric_type: MetricType.COSINE,
          params: JSON.stringify({ nprobe: 16 }),
        },
        output_fields: [
          'id', 'repo', 'path', 'content', 'chunk_type', 
          'language', 'start_line', 'end_line', 'star_count',
          'last_commit_date', 'text_hash', 'token_count'
        ],
        expr: this.buildFilterExpression(query),
      };

      const results = await this.client.search(searchParams);
      
      return this.formatSearchResults(results.results);
    } catch (error) {
      logger.error('Error searching embeddings:', error);
      throw error;
    }
  }

  async searchByText(
    query: SearchQuery,
    topK: number = 10
  ): Promise<SearchResult[]> {
    if (!this.isConnected) {
      await this.connect();
    }

    try {
      const collectionName = this.getCollectionName(query.chunkType);
      
      // Use text-based search (if available) or hybrid search
      const searchParams = {
        collection_name: collectionName,
        expr: this.buildTextSearchExpression(query),
        output_fields: [
          'id', 'repo', 'path', 'content', 'chunk_type', 
          'language', 'start_line', 'end_line', 'star_count',
          'last_commit_date', 'text_hash', 'token_count'
        ],
        limit: topK,
      };

      const results = await this.client.query(searchParams);
      
      return this.formatQueryResults(results.data);
    } catch (error) {
      logger.error('Error searching by text:', error);
      throw error;
    }
  }

  async getRepositoryFiles(repoName: string): Promise<SearchResult[]> {
    if (!this.isConnected) {
      await this.connect();
    }

    try {
      const searchParams = {
        collection_name: 'codex7_chunks',
        expr: `repo == "${repoName}"`,
        output_fields: [
          'id', 'repo', 'path', 'content', 'chunk_type', 
          'language', 'start_line', 'end_line', 'star_count',
          'last_commit_date', 'text_hash', 'token_count'
        ],
        limit: 1000, // Large limit to get all files
      };

      const results = await this.client.query(searchParams);
      
      return this.formatQueryResults(results.data);
    } catch (error) {
      logger.error('Error getting repository files:', error);
      throw error;
    }
  }

  async searchSymbols(symbolName: string, repo?: string): Promise<SearchResult[]> {
    if (!this.isConnected) {
      await this.connect();
    }

    try {
      let expr = `content like "%${symbolName}%"`;
      if (repo) {
        expr += ` and repo == "${repo}"`;
      }
      expr += ` and chunk_type == "code"`;

      const searchParams = {
        collection_name: 'codex7_chunks',
        expr,
        output_fields: [
          'id', 'repo', 'path', 'content', 'chunk_type', 
          'language', 'start_line', 'end_line', 'star_count',
          'last_commit_date', 'text_hash', 'token_count'
        ],
        limit: 50,
      };

      const results = await this.client.query(searchParams);
      
      return this.formatQueryResults(results.data);
    } catch (error) {
      logger.error('Error searching symbols:', error);
      throw error;
    }
  }

  async insertVectors(vectors: any[]): Promise<void> {
    if (!this.isConnected) {
      await this.connect();
    }

    try {
      if (vectors.length === 0) {
        logger.warn('No vectors to insert');
        return;
      }

      const collectionName = 'codex7_chunks';
      
      // Prepare data for insertion
      const data = vectors.map(vector => ({
        id: vector.id,
        vector: vector.vector,
        repo: vector.metadata.repo,
        path: vector.metadata.path,
        content: vector.content || '',
        chunk_type: vector.metadata.chunkType,
        language: vector.metadata.language,
        start_line: vector.metadata.startLine,
        end_line: vector.metadata.endLine,
        chunk_index: vector.metadata.chunkIndex,
        star_count: vector.metadata.starCount || 0,
        last_commit_date: vector.metadata.lastCommitDate || '',
        text_hash: vector.metadata.textHash || '',
        token_count: vector.metadata.tokenCount,
        content_length: vector.metadata.contentLength
      }));

      const insertParams = {
        collection_name: collectionName,
        data: data
      };

      await this.client.insert(insertParams);
      logger.info(`Inserted ${vectors.length} vectors into ${collectionName}`);
    } catch (error) {
      logger.error('Error inserting vectors:', error);
      throw error;
    }
  }

  async getCollectionStats(): Promise<any> {
    if (!this.isConnected) {
      await this.connect();
    }

    try {
      const stats = await this.client.getCollectionStatistics({
        collection_name: 'codex7_chunks',
      });
      
      return stats;
    } catch (error) {
      logger.error('Error getting collection stats:', error);
      throw error;
    }
  }

  private getCollectionName(chunkType: string | undefined): string {
    if (chunkType === 'doc') {
      return 'codex7_docs';
    } else if (chunkType === 'code') {
      return 'codex7_code';
    } else {
      return 'codex7_chunks'; // Default collection with both types
    }
  }

  private buildFilterExpression(query: SearchQuery): string {
    const filters: string[] = [];

    if (query.language) {
      filters.push(`language == "${query.language}"`);
    }

    if (query.repo) {
      filters.push(`repo == "${query.repo}"`);
    }

    if (query.chunkType && query.chunkType !== 'both') {
      filters.push(`chunk_type == "${query.chunkType}"`);
    }

    return filters.length > 0 ? filters.join(' and ') : '';
  }

  private buildTextSearchExpression(query: SearchQuery): string {
    const filters: string[] = [];

    // Add text search condition
    filters.push(`content like "%${query.query}%"`);

    if (query.language) {
      filters.push(`language == "${query.language}"`);
    }

    if (query.repo) {
      filters.push(`repo == "${query.repo}"`);
    }

    if (query.chunkType && query.chunkType !== 'both') {
      filters.push(`chunk_type == "${query.chunkType}"`);
    }

    return filters.join(' and ');
  }

  private formatSearchResults(results: any[]): SearchResult[] {
    return results.map(result => {
      const metadata: VectorMetadata = {
        repo: result.repo,
        path: result.path,
        chunkType: result.chunk_type,
        language: result.language,
        startLine: result.start_line,
        endLine: result.end_line,
        chunkIndex: result.chunk_index || 0,
        starCount: result.star_count,
        lastCommitDate: result.last_commit_date,
        textHash: result.text_hash,
        tokenCount: result.token_count || 0,
        contentLength: result.content_length || result.content?.length || 0,
      };

      return {
        id: result.id,
        repo: result.repo,
        path: result.path,
        content: result.content,
        score: result.score || 0,
        chunkType: result.chunk_type,
        language: result.language,
        startLine: result.start_line,
        endLine: result.end_line,
        metadata,
      };
    });
  }

  private formatQueryResults(results: any[]): SearchResult[] {
    return results.map(result => {
      const metadata: VectorMetadata = {
        repo: result.repo,
        path: result.path,
        chunkType: result.chunk_type,
        language: result.language,
        startLine: result.start_line,
        endLine: result.end_line,
        chunkIndex: result.chunk_index || 0,
        starCount: result.star_count,
        lastCommitDate: result.last_commit_date,
        textHash: result.text_hash,
        tokenCount: result.token_count || 0,
        contentLength: result.content_length || result.content?.length || 0,
      };

      return {
        id: result.id,
        repo: result.repo,
        path: result.path,
        content: result.content,
        score: 1.0, // Default score for query results
        chunkType: result.chunk_type,
        language: result.language,
        startLine: result.start_line,
        endLine: result.end_line,
        metadata,
      };
    });
  }

  // Database creation logic removed - using default database only

  private async ensureCollectionExists(): Promise<void> {
    try {
      const collectionName = 'codex7_chunks';
      const exists = await this.client.hasCollection({
        collection_name: collectionName,
      });
      
      if (!exists.value) {
        await this.createCollection(collectionName);
        logger.info(`Created collection: ${collectionName}`);
      }
    } catch (error) {
      logger.error('Error ensuring collection exists:', error);
      throw error;
    }
  }

  private async createCollection(collectionName: string): Promise<void> {
    const schema = {
      collection_name: collectionName,
      fields: [
        {
          name: 'id',
          data_type: DataType.VarChar,
          is_primary_key: true,
          max_length: 256,
        },
        {
          name: 'vector',
          data_type: DataType.FloatVector,
          dim: 1536, // OpenAI text-embedding-3-small dimension
        },
        {
          name: 'repo',
          data_type: DataType.VarChar,
          max_length: 256,
        },
        {
          name: 'path',
          data_type: DataType.VarChar,
          max_length: 512,
        },
        {
          name: 'content',
          data_type: DataType.VarChar,
          max_length: 32768,
        },
        {
          name: 'chunk_type',
          data_type: DataType.VarChar,
          max_length: 32,
        },
        {
          name: 'language',
          data_type: DataType.VarChar,
          max_length: 32,
        },
        {
          name: 'start_line',
          data_type: DataType.Int32,
        },
        {
          name: 'end_line',
          data_type: DataType.Int32,
        },
        {
          name: 'chunk_index',
          data_type: DataType.Int32,
        },
        {
          name: 'star_count',
          data_type: DataType.Int32,
        },
        {
          name: 'last_commit_date',
          data_type: DataType.VarChar,
          max_length: 32,
        },
        {
          name: 'text_hash',
          data_type: DataType.VarChar,
          max_length: 64,
        },
        {
          name: 'token_count',
          data_type: DataType.Int32,
        },
        {
          name: 'content_length',
          data_type: DataType.Int32,
        },
      ],
    };

    await this.client.createCollection(schema);
    
    // Create index for vector field
    await this.client.createIndex({
      collection_name: collectionName,
      field_name: 'vector',
      index_type: 'IVF_FLAT',
      metric_type: MetricType.COSINE,
      params: { nlist: 128 },
    });

    // Load collection
    await this.client.loadCollection({
      collection_name: collectionName,
    });
  }

  /**
   * Check if a codebase already exists in Milvus
   */
  async codebaseExists(repo: string): Promise<boolean> {
    if (!this.isConnected) {
      await this.connect();
    }

    try {
      const searchParams = {
        collection_name: 'codex7_chunks',
        filter: `repo == "${repo}"`,
        output_fields: ['id'],
        limit: 1,
      };

      const results = await this.client.query(searchParams);
      return results.data.length > 0;
    } catch (error) {
      logger.error(`Error checking if codebase ${repo} exists in Milvus:`, error);
      return false;
    }
  }

  /**
   * Clean up specific codebase data from Milvus
   */
  async cleanupCodebase(repo: string): Promise<void> {
    if (!this.isConnected) {
      await this.connect();
    }

    try {
      logger.info(`Cleaning up Milvus data for codebase: ${repo}`);
      
      const deleteParams = {
        collection_name: 'codex7_chunks',
        filter: `repo == "${repo}"`,
      };

      const result = await this.client.delete(deleteParams);
      logger.info(`Successfully cleaned up Milvus data for codebase: ${repo}, deleted ${result.delete_cnt} records`);
    } catch (error) {
      logger.error(`Error cleaning up Milvus data for codebase ${repo}:`, error);
    }
  }

  /**
   * Get repository statistics from Milvus
   */
  async getRepositoryStats(repo: string): Promise<number> {
    if (!this.isConnected) {
      await this.connect();
    }

    try {
      const searchParams = {
        collection_name: 'codex7_chunks',
        filter: `repo == "${repo}"`,
        output_fields: ['id'],
        limit: 0, // Just get count
      };

      const results = await this.client.query(searchParams);
      return results.data.length;
    } catch (error) {
      logger.error(`Error getting repository stats for ${repo}:`, error);
      return 0;
    }
  }
} 

File: src/query/milvus_client.py
from typing import List, Dict, Any, Optional, Union
from pymilvus import (
    connections,
    utility,
    FieldSchema,
    CollectionSchema,
    DataType,
    Collection,
)
import numpy as np

from ..config import settings
from ..types import CodeChunk, SearchResult
from ..utils.logger import app_logger


class MilvusClient:
    """Milvus client for vector database operations."""
    
    def __init__(self):
        self.logger = app_logger.bind(component="milvus_client")
        self.collection = None
        self.dimension = settings.milvus_dimension
        self.collection_name = settings.milvus_collection_name
        
        self._connect()
        self._ensure_collection()
    
    def _connect(self):
        """Connect to Milvus server."""
        try:
            connections.connect(
                "default",
                host=settings.milvus_host,
                port=settings.milvus_port,
            )
            self.logger.info(f"Connected to Milvus at {settings.milvus_host}:{settings.milvus_port}")
        except Exception as e:
            self.logger.error(f"Failed to connect to Milvus: {e}")
            raise
    
    def _ensure_collection(self):
        """Ensure collection exists with proper schema."""
        try:
            if utility.has_collection(self.collection_name):
                self.collection = Collection(self.collection_name)
                self.logger.info(f"Using existing collection: {self.collection_name}")
            else:
                self._create_collection()
        except Exception as e:
            self.logger.error(f"Failed to ensure collection: {e}")
            raise
    
    def _create_collection(self):
        """Create collection with proper schema."""
        try:
            # Define schema
            fields = [
                FieldSchema(name="id", dtype=DataType.VARCHAR, max_length=64, is_primary=True),
                FieldSchema(name="file_path", dtype=DataType.VARCHAR, max_length=512),
                FieldSchema(name="content", dtype=DataType.VARCHAR, max_length=65535),
                FieldSchema(name="start_line", dtype=DataType.INT32),
                FieldSchema(name="end_line", dtype=DataType.INT32),
                FieldSchema(name="language", dtype=DataType.VARCHAR, max_length=32),
                FieldSchema(name="chunk_type", dtype=DataType.VARCHAR, max_length=64),
                FieldSchema(name="metadata", dtype=DataType.JSON),
                FieldSchema(name="embedding", dtype=DataType.FLOAT_VECTOR, dim=self.dimension),
            ]
            
            schema = CollectionSchema(fields=fields)
            self.collection = Collection(self.collection_name, schema)
            
            # Create index
            index_params = {
                "metric_type": "L2",
                "index_type": "IVF_FLAT",
                "params": {"nlist": 1024},
            }
            
            self.collection.create_index("embedding", index_params)
            self.logger.info(f"Created collection: {self.collection_name}")
            
        except Exception as e:
            self.logger.error(f"Failed to create collection: {e}")
            raise
    
    def insert_chunks(self, chunks: List[CodeChunk]) -> int:
        """Insert code chunks into Milvus."""
        if not chunks:
            return 0
        
        try:
            self.logger.info(f"Inserting {len(chunks)} chunks into Milvus")
            
            # Prepare data for insertion
            data = {
                "id": [chunk.id for chunk in chunks],
                "file_path": [chunk.file_path for chunk in chunks],
                "content": [chunk.content for chunk in chunks],
                "start_line": [chunk.start_line for chunk in chunks],
                "end_line": [chunk.end_line for chunk in chunks],
                "language": [chunk.language for chunk in chunks],
                "chunk_type": [chunk.chunk_type for chunk in chunks],
                "metadata": [chunk.metadata for chunk in chunks],
                "embedding": [chunk.embedding for chunk in chunks],
            }
            
            # Insert data
            insert_result = self.collection.insert([data[field] for field in data.keys()])
            
            # Flush to ensure data is persisted
            self.collection.flush()
            
            self.logger.info(f"Successfully inserted {len(chunks)} chunks")
            return len(chunks)
            
        except Exception as e:
            self.logger.error(f"Failed to insert chunks: {e}")
            raise
    
    def search_similar(
        self,
        query_embedding: List[float],
        top_k: int = 10,
        filter_expression: Optional[str] = None,
        metric_type: str = "L2"
    ) -> List[SearchResult]:
        """Search for similar chunks using vector similarity."""
        try:
            self.logger.info(f"Searching for similar chunks with top_k={top_k}")
            
            # Load collection into memory
            self.collection.load()
            
            # Prepare search parameters
            search_params = {
                "metric_type": metric_type,
                "params": {"nprobe": 10},
            }
            
            # Perform search
            results = self.collection.search(
                data=[query_embedding],
                anns_field="embedding",
                param=search_params,
                limit=top_k,
                expr=filter_expression,
                output_fields=["id", "file_path", "content", "start_line", "end_line", 
                             "language", "chunk_type", "metadata"],
            )
            
            # Convert results to SearchResult objects
            search_results = []
            for hits in results:
                for i, hit in enumerate(hits):
                    chunk = CodeChunk(
                        id=hit.id,
                        file_path=hit.entity.get("file_path"),
                        content=hit.entity.get("content"),
                        start_line=hit.entity.get("start_line"),
                        end_line=hit.entity.get("end_line"),
                        language=hit.entity.get("language"),
                        chunk_type=hit.entity.get("chunk_type"),
                        metadata=hit.entity.get("metadata", {}),
                    )
                    
                    search_result = SearchResult(
                        chunk=chunk,
                        score=hit.score,
                        rank=i + 1,
                        search_type="vector",
                        metadata={"distance": hit.distance},
                    )
                    search_results.append(search_result)
            
            self.logger.info(f"Found {len(search_results)} similar chunks")
            return search_results
            
        except Exception as e:
            self.logger.error(f"Failed to search similar chunks: {e}")
            raise
    
    def delete_by_file_path(self, file_path: str) -> int:
        """Delete all chunks for a specific file."""
        try:
            # Find chunks for the file
            expr = f'file_path == "{file_path}"'
            self.collection.delete(expr)
            
            self.logger.info(f"Deleted chunks for file: {file_path}")
            return 0  # Milvus doesn't return count
            
        except Exception as e:
            self.logger.error(f"Failed to delete chunks for file {file_path}: {e}")
            raise
    
    def get_chunk_by_id(self, chunk_id: str) -> Optional[CodeChunk]:
        """Get a specific chunk by ID."""
        try:
            expr = f'id == "{chunk_id}"'
            results = self.collection.query(
                expr=expr,
                output_fields=["id", "file_path", "content", "start_line", "end_line", 
                             "language", "chunk_type", "metadata", "embedding"],
            )
            
            if results:
                result = results[0]
                return CodeChunk(
                    id=result["id"],
                    file_path=result["file_path"],
                    content=result["content"],
                    start_line=result["start_line"],
                    end_line=result["end_line"],
                    language=result["language"],
                    chunk_type=result["chunk_type"],
                    metadata=result["metadata"],
                    embedding=result.get("embedding"),
                )
            
            return None
            
        except Exception as e:
            self.logger.error(f"Failed to get chunk {chunk_id}: {e}")
            return None
    
    def get_chunks_by_file(self, file_path: str) -> List[CodeChunk]:
        """Get all chunks for a specific file."""
        try:
            expr = f'file_path == "{file_path}"'
            results = self.collection.query(
                expr=expr,
                output_fields=["id", "file_path", "content", "start_line", "end_line", 
                             "language", "chunk_type", "metadata", "embedding"],
            )
            
            chunks = []
            for result in results:
                chunk = CodeChunk(
                    id=result["id"],
                    file_path=result["file_path"],
                    content=result["content"],
                    start_line=result["start_line"],
                    end_line=result["end_line"],
                    language=result["language"],
                    chunk_type=result["chunk_type"],
                    metadata=result["metadata"],
                    embedding=result.get("embedding"),
                )
                chunks.append(chunk)
            
            return chunks
            
        except Exception as e:
            self.logger.error(f"Failed to get chunks for file {file_path}: {e}")
            return []
    
    def drop_collection(self):
        """Drop the entire collection."""
        try:
            if utility.has_collection(self.collection_name):
                utility.drop_collection(self.collection_name)
                self.logger.info(f"Dropped collection: {self.collection_name}")
        except Exception as e:
            self.logger.error(f"Failed to drop collection: {e}")
            raise
    
    def get_collection_stats(self) -> Dict[str, Any]:
        """Get collection statistics."""
        try:
            if not self.collection:
                return {"error": "Collection not initialized"}
            
            stats = {
                "collection_name": self.collection_name,
                "num_entities": self.collection.num_entities,
            }
            
            return stats
            
        except Exception as e:
            self.logger.error(f"Failed to get collection stats: {e}")
            return {"error": str(e)}
    
    def close(self):
        """Close connection to Milvus."""
        try:
            connections.disconnect("default")
            self.logger.info("Disconnected from Milvus")
        except Exception as e:
            self.logger.error(f"Failed to disconnect from Milvus: {e}")